File: app/src/main/java/com/hhh/paws/ui/petProfile/menu/procedures/DetailProceduresFragment.java
package com.hhh.paws.ui.petProfile.menu.procedures;

import android.annotation.SuppressLint;
import android.app.DatePickerDialog;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavArgs;
import androidx.navigation.Navigation;

import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.DatePicker;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.google.android.material.textfield.MaterialAutoCompleteTextView;
import com.google.android.material.textfield.TextInputEditText;
import com.hhh.paws.R;
import com.hhh.paws.database.model.SurgicalProcedure;
import com.hhh.paws.database.viewModel.ProcedureViewModel;
import com.hhh.paws.databinding.FragmentProceduresDetailBinding;

import java.util.Calendar;
import java.util.UUID;

import dagger.hilt.android.AndroidEntryPoint;
import io.reactivex.rxjava3.disposables.Disposable;


@AndroidEntryPoint
public class DetailProceduresFragment extends Fragment implements DatePickerDialog.OnDateSetListener {

    private FragmentProceduresDetailBinding _binding = null;
    private FragmentProceduresDetailBinding getBinding() {
        return _binding;
    }

    private MaterialAutoCompleteTextView spinnerTypeSurgicalProcedure;
    private TextInputEditText nameSurgicalProcedure;
    private TextInputEditText anesthesiaSurgicalProcedure;
    private TextInputEditText dateSurgicalProcedure;
    private TextInputEditText veterinarianSurgicalProcedure;
    private TextInputEditText descriptionSurgicalProcedure;

    private int day = 0;
    private int month = 0;
    private int year = 0;
    private int savedDay = 0;
    private int savedMonth = 0;
    private int savedYear = 0;

    private ProcedureViewModel viewModelProcedures;
    private String petNameThis;
    private SurgicalProcedure procedure = null;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        _binding = FragmentProceduresDetailBinding.inflate(inflater, container, false);
        viewModelProcedures = new ViewModelProvider(
                DetailProceduresFragment.this).get(ProcedureViewModel.class

        );
        return getBinding().getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        petNameThis = "Котик";
        procedure = getArguments().getParcelable("procedure");

        spinnerTypeSurgicalProcedure = getBinding().spinnerTypeSurgicalProcedure;
        nameSurgicalProcedure = getBinding().nameSurgicalProcedure;
        anesthesiaSurgicalProcedure = getBinding().anesthesiaSurgicalProcedure;
        veterinarianSurgicalProcedure = getBinding().veterinarianSurgicalProcedure;
        descriptionSurgicalProcedure = getBinding().descriptionSurgicalProcedure;

        dateSurgicalProcedure = getBinding().dateSurgicalProcedure;
        dateSurgicalProcedure.setOnClickListener(v -> {
            getDateSet();
            new DatePickerDialog(requireContext(), this, year, month, day).show();
        });

        if (procedure != null) {
            spinnerTypeSurgicalProcedure.setText(procedure.getType());
            nameSurgicalProcedure.setText(procedure.getName());
            anesthesiaSurgicalProcedure.setText(procedure.getAnesthesia());
            dateSurgicalProcedure.setText(procedure.getDate());
            veterinarianSurgicalProcedure.setText(procedure.getVeterinarian());
            descriptionSurgicalProcedure.setText(procedure.getDescription());
        }
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.toolbar_fragment_vetpassport, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.action_save) {
            SurgicalProcedure newProcedure = new SurgicalProcedure();
            if (procedure != null) {
                newProcedure.setId(procedure.getId());
            } else {
                newProcedure.setId(UUID.randomUUID().toString());
            }
            newProcedure.setAnesthesia(anesthesiaSurgicalProcedure.getText().toString());
            newProcedure.setDate(dateSurgicalProcedure.getText().toString());
            newProcedure.setDescription(descriptionSurgicalProcedure.getText().toString());
            newProcedure.setVeterinarian(veterinarianSurgicalProcedure.getText().toString());
            newProcedure.setType(spinnerTypeSurgicalProcedure.getText().toString());
            newProcedure.setName(nameSurgicalProcedure.getText().toString());

            boolean flag = viewModelProcedures.setProcedure(petNameThis, newProcedure);
            if (flag) {
                Navigation.findNavController(requireActivity(), R.id.nav_host_fragment_content_vet_passport)
                        .popBackStack();
            } else {
                Toast.makeText(requireContext(), "error", Toast.LENGTH_SHORT).show();
            }

            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        _binding = null;
    }

    private void getDateSet() {
        day = Calendar.DAY_OF_WEEK_IN_MONTH;
        month = Calendar.MONTH;
        year = Calendar.YEAR;
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
        savedDay = dayOfMonth;
        savedMonth = month + 1;
        savedYear = year;

        dateSurgicalProcedure.setText(savedDay + "." + savedMonth + "." + savedYear);
    }
}

File: app/src/main/java/com/hhh/paws/ui/petProfile/menu/procedures/DetailProceduresFragment.kt
package com.hhh.paws.ui.petProfile.menu.procedures

import android.annotation.SuppressLint
import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.os.Bundle
import android.view.*
import android.widget.DatePicker
import android.widget.ProgressBar
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import androidx.navigation.Navigation.findNavController
import androidx.navigation.fragment.navArgs
import com.google.android.material.textfield.MaterialAutoCompleteTextView
import com.google.android.material.textfield.TextInputEditText
import com.hhh.paws.R
import com.hhh.paws.database.model.SurgicalProcedure
import com.hhh.paws.database.viewModel.ProcedureViewModel
import com.hhh.paws.databinding.FragmentDetailProceduresBinding
import com.hhh.paws.ui.petProfile.menu.notes.DetailNoteFragmentArgs
import com.hhh.paws.util.PetName.name
import com.hhh.paws.util.UiState
import dagger.hilt.android.AndroidEntryPoint
import java.util.*


@AndroidEntryPoint
class DetailProceduresFragment : Fragment(), OnDateSetListener {

    private var binding: FragmentDetailProceduresBinding? = null
    private var spinnerTypeSurgicalProcedure: MaterialAutoCompleteTextView? = null
    private var nameSurgicalProcedure: TextInputEditText? = null
    private var anesthesiaSurgicalProcedure: TextInputEditText? = null
    private var dateSurgicalProcedure: TextInputEditText? = null
    private var veterinarianSurgicalProcedure: TextInputEditText? = null
    private var descriptionSurgicalProcedure: TextInputEditText? = null
    private var progressBarProcedureDetail: ProgressBar? = null
    private var day = 0
    private var month = 0
    private var year = 0
    private var savedDay = 0
    private var savedMonth = 0
    private var savedYear = 0
    private val viewModelProcedures by viewModels<ProcedureViewModel>()
    private val bundleArgs: DetailProceduresFragmentArgs by navArgs()
    private var petNameThis: String? = null
    private var procedure: SurgicalProcedure? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentDetailProceduresBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        petNameThis = name
        procedure = bundleArgs.procedure

        progressBarProcedureDetail = binding!!.progressBarProcedureDetail
        spinnerTypeSurgicalProcedure = binding!!.spinnerTypeSurgicalProcedure
        nameSurgicalProcedure = binding!!.nameSurgicalProcedure
        anesthesiaSurgicalProcedure = binding!!.anesthesiaSurgicalProcedure
        veterinarianSurgicalProcedure = binding!!.veterinarianSurgicalProcedure
        descriptionSurgicalProcedure = binding!!.descriptionSurgicalProcedure
        dateSurgicalProcedure = binding!!.dateSurgicalProcedure

        dateSurgicalProcedure!!.setOnClickListener { v: View? ->
            dateSet
            DatePickerDialog(requireContext(), this, year, month, day).show()
        }

        if (procedure != null) {
            spinnerTypeSurgicalProcedure!!.setText(procedure!!.type)
            nameSurgicalProcedure!!.setText(procedure!!.name)
            anesthesiaSurgicalProcedure!!.setText(procedure!!.anesthesia)
            dateSurgicalProcedure!!.setText(procedure!!.date)
            veterinarianSurgicalProcedure!!.setText(procedure!!.veterinarian)
            descriptionSurgicalProcedure!!.setText(procedure!!.description)
        }

        viewModelProcedures.addProcedure.observe(viewLifecycleOwner) {
            when (it) {
                is UiState.Loading -> {
                    progressBarProcedureDetail?.visibility = View.VISIBLE
                }
                is UiState.Success -> {
                    progressBarProcedureDetail?.visibility = View.INVISIBLE
                    findNavController(requireActivity(), R.id.nav_host_fragment_content_vet_passport)
                        .popBackStack()
                }
                is UiState.Failure -> {
                    progressBarProcedureDetail?.visibility = View.INVISIBLE
                }
                else -> {
                    progressBarProcedureDetail?.visibility = View.INVISIBLE
                }
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.toolbar_fragment_vetpassport, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return if (item.itemId == R.id.action_save) {
            progressBarProcedureDetail!!.visibility = View.VISIBLE
            val newProcedure = SurgicalProcedure()
            if (procedure != null) {
                newProcedure.id = procedure!!.id
            } else {
                newProcedure.id = UUID.randomUUID().toString()
            }
            newProcedure.anesthesia = anesthesiaSurgicalProcedure!!.text.toString()
            newProcedure.date = dateSurgicalProcedure!!.text.toString()
            newProcedure.description = descriptionSurgicalProcedure!!.text.toString()
            newProcedure.veterinarian = veterinarianSurgicalProcedure!!.text.toString()
            newProcedure.type = spinnerTypeSurgicalProcedure!!.text.toString()
            newProcedure.name = nameSurgicalProcedure!!.text.toString()
            progressBarProcedureDetail!!.visibility = View.INVISIBLE
            viewModelProcedures.setProcedure(petNameThis!!, newProcedure)
            findNavController(requireActivity(), R.id.nav_host_fragment_content_vet_passport)
                .popBackStack()
            true
        } else {
            super.onOptionsItemSelected(item)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        binding = null
    }

    private val dateSet: Unit
        get() {
            val calendar = Calendar.getInstance()
            year = calendar[Calendar.YEAR]
            month = calendar[Calendar.MONTH]
            day = calendar[Calendar.DAY_OF_MONTH]
        }

    @SuppressLint("SetTextI18n")
    override fun onDateSet(view: DatePicker, year: Int, month: Int, dayOfMonth: Int) {
        savedDay = dayOfMonth
        savedMonth = month + 1
        savedYear = year
        dateSurgicalProcedure!!.setText("$savedDay.$savedMonth.$savedYear")
    }
}